x-pack/plugin/downsample/src/main/java/org/elasticsearch/xpack/downsample/DownsampleShardPersistentTaskExecutor.java

```diff
@@ -110,6 +121,16 @@ public PersistentTasksCustomMetadata.Assignment getAssignment(
         // Here we make sure we assign the task to the actual node holding the shard identified by
         // the downsampling task shard id.
         final ShardId shardId = params.shardId();
+
+        // If during re-assignment the source index disappeared, then we need to break out.
+        // Returning NO_NODE_FOUND just keeps the persistent task until the source index appears again (which would never happen)
+        // So let's return a node and then in the node operation we would just fail and stop this persistent task
+        var indexShardRouting = clusterState.routingTable().shardRoutingTable(params.shardId());
```

Comment:
I don't understand why we do this "before" the next if block.

Wouldn't it be better to do it inside the if block if we do not find any node? My understading is that if we try to find a node and we happen not to find any then it could be because the index is not there anymore.

Then maybe we can just add a check into the Downsample task which checks for the existence of the source index at the very beginning and fails with an "unrecoverable" error. In that case the executor would not retry right?

Reply:
Or maybe we don't even need this logic at all...

in the Downsample task, at the very beginning when fetching the initial state we could check if the index exists. If not we can fail with an "unrecoverable" error. That should stop the executor from indefinitely retrying right?

Reply:
Wait I see what is the issue here...if the index is deleted there is no shard with index data -> there is no shard to start a task and we are not able to start a downsample task to check the existence of the index...

Reply:
> I don't understand why we do this "before" the next if block.

This is because with: `final ShardRouting shardRouting = clusterState.routingTable().shardRoutingTable(shardId).primaryShard();`, this part `shardRoutingTable(shardId)` will fail. This is why I moved this up.

Reply:
> Wait I see what is the issue here...if the index is deleted there is no shard with index data -> there is no shard to start a task and we are not able to start a downsample task to check the existence of the index...

Yes, we're basically stuck forever. The getting an assignment fails with an exception and the persistent taks framework tries again on next cluster state update. But it will keep on failing because the index has been removed.